File: 41. First Missing Positive/Solution.py
# LeetCode #41: First Missing Positive

'''
Given an unsorted integer array, find the first missing positive integer.

For example,
Given [1,2,0] return 3,
and [3,4,-1,1] return 2.

Your algorithm should run in O(n) time and uses constant space.
'''

class Solution:
    # @param {integer[]} nums
    # @return {integer}
    def firstMissingPositive(self, nums):
        i = 0
        n = len(nums)
        
        while i < n:
            if nums[i] != i+1 and nums[i] > 0 and nums[i] <= n and nums[i] != nums[nums[i]-1]:
                self.swap(nums, i, nums[i]-1)
            else:
                i += 1
                
        for j in xrange(n):
            if nums[j] != j+1:
                return j+1
                
        return n+1
        
    def swap(self, nums, i, j):
        temp = nums[i]
        nums[i] = nums[j]
        nums[j] = temp
            

File: 41. First Missing Positive/Solution.java
// LeetCode #41: First Missing Positive

/*
Given an unsorted integer array, find the first missing positive integer.

For example,
Given [1,2,0] return 3,
and [3,4,-1,1] return 2.

Your algorithm should run in O(n) time and uses constant space.
*/

public class Solution {
    public int firstMissingPositive(int[] nums) {
        int i = 0;
        while (i < nums.length) {
            if (nums[i] != i+1 && nums[i] > 0 && nums[i] <= nums.length && nums[i] != nums[nums[i] - 1])
                swap(nums, i, nums[i]-1);
            else
                i++;
        }
        
        for (int j = 0; j < nums.length; j++) {
            if (nums[j] != j+1)
                return j+1;
        }
        return nums.length+1;
    }
    
    private void swap(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
}